File: backend/app/domains/ab_testing/schemas/ab_testing.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator


class CreateAbExperimentRequest(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    description: str | None = Field(default=None, max_length=8000)
    evaluation_set_id: str
    metrics_config: dict[str, object] = Field(default_factory=dict)

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: str) -> str:
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("name must not be blank")
        return trimmed

    @field_validator("evaluation_set_id")
    @classmethod
    def validate_evaluation_set_id(cls, value: str) -> str:
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("evaluation_set_id must not be blank")
        return trimmed


class UpdateAbExperimentRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    description: str | None = None
    metrics_config: dict[str, object] | None = None

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: str | None) -> str | None:
        if value is None:
            return value
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("name must not be blank")
        return trimmed


class CreateAbVariantRequest(BaseModel):
    label: str = Field(min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=4000)
    rag_profile_id: str | None = None
    rag_profile_version: int | None = Field(default=None, ge=1)
    prompt_template_version_id: str | None = None
    model_profile_key: str | None = Field(default=None, max_length=64)
    config_snapshot: dict[str, object] = Field(default_factory=dict)

    @field_validator("label")
    @classmethod
    def validate_label(cls, value: str) -> str:
        trimmed = value.strip()
        if not trimmed:
            raise ValueError("label must not be blank")
        return trimmed


class AbVariantResponse(BaseModel):
    variant_id: str
    experiment_id: str
    label: str
    description: str | None = None
    rag_profile_id: str | None = None
    rag_profile_version: int | None = None
    prompt_template_version_id: str | None = None
    model_profile_key: str | None = None
    config_snapshot: dict[str, object]
    approval_status: Literal["pending", "approved", "rejected"]
    approved_by_id: str | None = None
    approval_note: str | None = None
    approved_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class AbExperimentResponse(BaseModel):
    experiment_id: str
    name: str
    description: str | None = None
    evaluation_set_id: str
    status: Literal["draft", "running", "completed", "failed"]
    metrics_config: dict[str, object]
    created_by_id: str | None = None
    created_at: datetime
    updated_at: datetime
    variants: list[AbVariantResponse] = Field(default_factory=list)


class AbExperimentListResponse(BaseModel):
    items: list[AbExperimentResponse]
    total: int
    limit: int
    offset: int


class StartAbExperimentRunRequest(BaseModel):
    """Trigger a new run of the experiment against all current variants."""

    note: str | None = Field(default=None, max_length=2000)


# ---------------------------------------------------------------------------
# Variant run / comparison
# ---------------------------------------------------------------------------


class VariantMetricDelta(BaseModel):
    """Per-metric comparison between a variant and a reference (first variant or control)."""

    metric: str
    label: str
    reference_value: float | None
    variant_value: float | None
    delta: float | None
    improved: bool | None


class VariantRunSummary(BaseModel):
    variant_id: str
    variant_label: str
    evaluation_run_id: str | None = None
    status: str
    metrics_summary: dict[str, object]
    deltas_vs_reference: list[VariantMetricDelta] = Field(default_factory=list)
    error_detail: str | None = None


class AbExperimentRunResponse(BaseModel):
    experiment_run_id: str
    experiment_id: str
    status: Literal["draft", "running", "completed", "failed"]
    triggered_by_id: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    variant_summaries: list[VariantRunSummary] = Field(default_factory=list)
    comparison_report: dict[str, object]


class AbExperimentRunListResponse(BaseModel):
    items: list[AbExperimentRunResponse]
    total: int
    limit: int
    offset: int


# ---------------------------------------------------------------------------
# Approval
# ---------------------------------------------------------------------------


class ApproveVariantRequest(BaseModel):
    """Promote a variant's RAG profile to the org default after experiment approval."""

    note: str | None = Field(default=None, max_length=2000)
    set_as_default_profile: bool = Field(
        default=False,
        description=(
            "When true, sets the variant's rag_profile as the org default. "
            "Requires the experiment run to be completed."
        ),
    )


class RejectVariantRequest(BaseModel):
    note: str | None = Field(default=None, max_length=2000)

File: frontend/src/lib/api/ab-testing.ts
import { apiRequest } from "@/lib/api/request";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export type AbExperimentStatus = "draft" | "running" | "completed" | "failed";
export type AbVariantApprovalStatus = "pending" | "approved" | "rejected";

export type AbVariantResponse = {
  variant_id: string;
  experiment_id: string;
  label: string;
  description: string | null;
  rag_profile_id: string | null;
  rag_profile_version: number | null;
  prompt_template_version_id: string | null;
  model_profile_key: string | null;
  config_snapshot: Record<string, unknown>;
  approval_status: AbVariantApprovalStatus;
  approved_by_id: string | null;
  approval_note: string | null;
  approved_at: string | null;
  created_at: string;
  updated_at: string;
};

export type AbExperimentResponse = {
  experiment_id: string;
  name: string;
  description: string | null;
  evaluation_set_id: string;
  status: AbExperimentStatus;
  metrics_config: Record<string, unknown>;
  created_by_id: string | null;
  created_at: string;
  updated_at: string;
  variants: AbVariantResponse[];
};

export type AbExperimentListResponse = {
  items: AbExperimentResponse[];
  total: number;
  limit: number;
  offset: number;
};

export type VariantMetricDelta = {
  metric: string;
  label: string;
  reference_value: number | null;
  variant_value: number | null;
  delta: number | null;
  improved: boolean | null;
};

export type VariantRunSummary = {
  variant_id: string;
  variant_label: string;
  evaluation_run_id: string | null;
  status: string;
  metrics_summary: Record<string, number | null>;
  deltas_vs_reference: VariantMetricDelta[];
  error_detail: string | null;
};

export type AbExperimentRunResponse = {
  experiment_run_id: string;
  experiment_id: string;
  status: AbExperimentStatus;
  triggered_by_id: string | null;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
  variant_summaries: VariantRunSummary[];
  comparison_report: Record<string, unknown>;
};

export type AbExperimentRunListResponse = {
  items: AbExperimentRunResponse[];
  total: number;
  limit: number;
  offset: number;
};

// ---------------------------------------------------------------------------
// Request payloads
// ---------------------------------------------------------------------------

export type CreateAbExperimentRequest = {
  name: string;
  description?: string | null;
  evaluation_set_id: string;
  metrics_config?: Record<string, unknown>;
};

export type UpdateAbExperimentRequest = {
  name?: string | null;
  description?: string | null;
  metrics_config?: Record<string, unknown> | null;
};

export type CreateAbVariantRequest = {
  label: string;
  description?: string | null;
  rag_profile_id?: string | null;
  rag_profile_version?: number | null;
  prompt_template_version_id?: string | null;
  model_profile_key?: string | null;
  config_snapshot?: Record<string, unknown>;
};

export type StartAbExperimentRunRequest = {
  note?: string | null;
};

export type ApproveVariantRequest = {
  note?: string | null;
  set_as_default_profile?: boolean;
};

export type RejectVariantRequest = {
  note?: string | null;
};

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

const BASE = "/ab-experiments";

export async function listAbExperiments(params?: {
  limit?: number;
  offset?: number;
}): Promise<AbExperimentListResponse> {
  const qs = new URLSearchParams();
  if (params?.limit != null) qs.set("limit", String(params.limit));
  if (params?.offset != null) qs.set("offset", String(params.offset));
  return apiRequest<AbExperimentListResponse>(`${BASE}?${qs}`);
}

export async function createAbExperiment(
  payload: CreateAbExperimentRequest
): Promise<AbExperimentResponse> {
  return apiRequest<AbExperimentResponse>(BASE, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function getAbExperiment(experimentId: string): Promise<AbExperimentResponse> {
  return apiRequest<AbExperimentResponse>(`${BASE}/${experimentId}`);
}

export async function updateAbExperiment(
  experimentId: string,
  payload: UpdateAbExperimentRequest
): Promise<AbExperimentResponse> {
  return apiRequest<AbExperimentResponse>(`${BASE}/${experimentId}`, {
    method: "PATCH",
    body: JSON.stringify(payload),
  });
}

export async function deleteAbExperiment(experimentId: string): Promise<void> {
  await apiRequest<void>(`${BASE}/${experimentId}`, { method: "DELETE" });
}

export async function addVariant(
  experimentId: string,
  payload: CreateAbVariantRequest
): Promise<AbVariantResponse> {
  return apiRequest<AbVariantResponse>(`${BASE}/${experimentId}/variants`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function removeVariant(experimentId: string, variantId: string): Promise<void> {
  await apiRequest<void>(`${BASE}/${experimentId}/variants/${variantId}`, { method: "DELETE" });
}

export async function startExperimentRun(
  experimentId: string,
  payload?: StartAbExperimentRunRequest
): Promise<AbExperimentRunResponse> {
  return apiRequest<AbExperimentRunResponse>(`${BASE}/${experimentId}/runs`, {
    method: "POST",
    body: JSON.stringify(payload ?? {}),
  });
}

export async function listExperimentRuns(
  experimentId: string,
  params?: { limit?: number; offset?: number }
): Promise<AbExperimentRunListResponse> {
  const qs = new URLSearchParams();
  if (params?.limit != null) qs.set("limit", String(params.limit));
  if (params?.offset != null) qs.set("offset", String(params.offset));
  return apiRequest<AbExperimentRunListResponse>(
    `${BASE}/${experimentId}/runs?${qs}`
  );
}

export async function getExperimentRun(
  experimentId: string,
  runId: string
): Promise<AbExperimentRunResponse> {
  return apiRequest<AbExperimentRunResponse>(`${BASE}/${experimentId}/runs/${runId}`);
}

export async function finalizeExperimentRun(
  experimentId: string,
  runId: string
): Promise<AbExperimentRunResponse> {
  return apiRequest<AbExperimentRunResponse>(
    `${BASE}/${experimentId}/runs/${runId}/finalize`,
    { method: "POST", body: JSON.stringify({}) }
  );
}

export async function approveVariant(
  experimentId: string,
  variantId: string,
  payload?: ApproveVariantRequest
): Promise<AbVariantResponse> {
  return apiRequest<AbVariantResponse>(
    `${BASE}/${experimentId}/variants/${variantId}/approve`,
    { method: "POST", body: JSON.stringify(payload ?? {}) }
  );
}

export async function rejectVariant(
  experimentId: string,
  variantId: string,
  payload?: RejectVariantRequest
): Promise<AbVariantResponse> {
  return apiRequest<AbVariantResponse>(
    `${BASE}/${experimentId}/variants/${variantId}/reject`,
    { method: "POST", body: JSON.stringify(payload ?? {}) }
  );
}
